### PR TITLE
Tweak completion filtering behavior.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems?id=362890

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide .

**Comment by DustinCampbell** [comment]
cc @mattwar for API changes

File: src/Features/Core/Portable/Completion/CompletionItem.cs [hunk 1]
```diff
@@ -257,9 +257,33 @@ int IComparable<CompletionItem>.CompareTo(CompletionItem other)
             return result;
         }
 
-        public override string ToString()
+        public override string ToString() => DisplayText;
+
+        internal bool ShouldBeFilteredOutOfCompletionList(
+            ImmutableDictionary<CompletionItemFilter, bool> filterState)
```
Comment (mattwar) [inline src/Features/Core/Portable/Completion/CompletionItem.cs]: Does this function belong on this type?  It doesn't use any of the class's state.

File: src/Features/Core/Portable/Completion/CompletionTrigger.cs [hunk 2]
```diff
@@ -28,22 +29,19 @@ internal CompletionTrigger(CompletionTriggerKind kind, char character = (char)0)
         /// <summary>
         /// The default <see cref="CompletionTrigger"/> when none is specified.
         /// </summary>
+        [Obsolete("Do not use.", error: true)]
```
Comment (mattwar) [inline src/Features/Core/Portable/Completion/CompletionTrigger.cs]: >[Obsolete("Do not use.", error: true)] [](start = 8, length = 38)

It seems bizarre to have a Default that should not be used.  Should the default just be changed to default to Invoke instead?

File: src/Features/Core/Portable/Completion/CompletionItem.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionItem.cs]: I'll find a good home for it.

File: src/Features/Core/Portable/Completion/CompletionTrigger.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionTrigger.cs]: I can do that.  part o the reason i wanted to make this change was to force people to have to think about what they were doing.  But i think it's acceptably reasonable to have hte 'Default' and having it be set to 'Invoke'.

**Comment by CyrusNajmabadi** [comment]
Ok.  Have responded to PR feedback.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/CompletionFilterReason.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    internal enum CompletionFilterReason
+    {
+        Insertion,
+        Deletion,
+        NonInsertionOrDeletion,
+#if false
+        // If necessary, we could add additional filter reasons.  For example, for the below items.
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/CompletionFilterReason.cs]: This comment feels much like "I'm commenting out code I thought I needed". Perhaps rewrite to more clear prose?

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs [hunk 4]
```diff
@@ -170,11 +146,23 @@ internal partial class Session
...
+                        // if one of two things is true:
+                        //
+                        //  1. The user has only typed a single character.  In this case they might
+                        //     have just typed the character to get completion.  Filtering out items
+                        //     here is not desirable.
+                        //
+                        //  2. They brough up completion with ctrl-j or through deletion.  In these
+                        //     cases we just always keep all the items in the list.
+
+                        var wasTriggeredByDeleteOrSimpleInvoke =
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs]: Reorder these to match the cases above.

Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs]: But massive thanks for actually explaining in English what the heck this is trying to do.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/CompletionFilterReason.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    internal enum CompletionFilterReason
+    {
+        Insertion,
+        Deletion,
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/CompletionFilterReason.cs]: Please add a doc comment explaining this is either delete or backspace.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/FilterResult.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+using Microsoft.CodeAnalysis.Completion;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion
+{
+    internal partial class Controller
+    {
+        internal partial class Session
+        {
+            private struct FilterResult
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/FilterResult.cs]: File name here isn't matching the nesting structure.

**Pull request merged**